Admin-side navbar collapse

seand7565 opened the pull request:
**Description**
Makes the admin-side navbar collapsible by clicking a collapse icon at the top of the page.

![navbar-collapse](https://user-images.githubusercontent.com/5720486/55765440-5659c100-5a35-11e9-9669-77df1564866d.gif)

Additional functionality:

- The navbar defaults to closed on iPad sized screens or smaller.
- When the navbar is collapsed or opened, a cookie is stored that remembers your preference - on the next page load, the navbar will be in the same state as it was on the previous page.
- When the navbar is closed, a tooltip is set for the navbar icons which - on hover- displays the name of the button you have selected. If you prefer a closed navbar, but don't know what the icons mean, you can still find your way around.

![image](https://user-images.githubusercontent.com/5720486/55765529-d8e28080-5a35-11e9-9769-4b5e727ff84c.png)

@jacobherrington and @tvdeyen did most of the heavy lifting here, I just cleaned it up and added a few things.

ref #2961 - should close this issue if everything looks good!

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [ ] I have added tests to cover this change (if needed)

seand7565 commented:
Got some tests to fix here - capybara is trying to click on nested navbar links, but those are hidden when the navbar is collapsed. The navbar defaults to closed on small screens, which is the problem for these tests. 

The two solutions I have are to increase capybaras window size, or just click the navbar-toggle button to open the navbar before running the tests. I'll pop something in soon to get the tests to pass.

tvdeyen commented:
Thanks, this looks great! Regarding the specs: My vote is to raise the capybara window size.

aitbw commented:
My vote too goes to raise Capybara window size. This looks very good, @seand7565! Thanks a lot :raised_hands:

One small thing: is it possible you could rebase this PR? That'd would help us avoid the merge commit in here

seand7565 commented:
(Mostly) passing now! @aitbw The screen size actually wasn't the issue here like I thought, I used the wrong operator in the javascript file determining the default screen size (greater than rather than less than). D'oh!

Not sure why the mysql rails 5.1 tests failed - and on the flash too, which wasn't touched in this commit. I'll see if I can find out why it failed!

Thanks all for the feedback! ❤️ I rebased and squashed some of the smaller changes - not sure how far you folks like PRs squashed. Should this all be one commit?

seand7565 commented:
The test that was previously failing here appears to be passing now after a rebase 🤷‍♂

spaghetticode commented:
@seand7565 Thanks for the PR 👏 

I see it now needs some rebasing, some asset files are out-of-sync, can you please update them? Thank you!

seand7565 commented:
This doesn't play well with the new admin styling - will have to make some changes to get it back in line. Give me a few days and I'll have it done!

seand7565 commented:
Should be good to go now!

![image](https://user-images.githubusercontent.com/5720486/59237574-62dad100-8bc0-11e9-8f95-b83270a59757.png)

![image](https://user-images.githubusercontent.com/5720486/59237581-6706ee80-8bc0-11e9-80f6-369314475aeb.png)

elia commented on backend/app/assets/javascripts/spree/backend/components/admin_nav.js:
```diff
@@ -0,0 +1,27 @@
+  if (window.screen.width <= 1024 && !document.cookie.includes("admin_nav_hidden")) {
+    // Set default nav to collapse on small screens - but don't override user preference
+    document.body.classList.add("admin-nav-hidden");
+    document.cookie = "admin_nav_hidden=true; expires=Fri, 31 Dec 9999 23:59:59 GMT";
```
Any reason not to use LocalStorage instead? 
Browser support is pretty wide https://caniuse.com/#feat=namevalue-storage
I would also suggest namespacing the key with `solidus.` or `solidus/`, e.g. `window.localStorage.get('solidus.minimize-admin-nav')`

jacobherrington replied:
I believe I was initially using LocalStorage, but there was some feedback on that decision from either @tvdeyen or @seand7565. I do not remember it 🤦‍♂

kennyadsl commented:
I'm ok with this but I think it will need some extra design iterations after the merge. If you agree I'd merge it after v2.9 release which already contains a lot of backend changes.

mfrecchiami commented:
Hi @seand7565 and everyone!

I tried to improve a little bit more this new feature, moving the collapse icon inside the `admin-nav-header` and leaving the Solidus mark always visible.

![minimize-menu](https://user-images.githubusercontent.com/46188345/61550341-7ba59480-aa52-11e9-9b9b-e6ce01e4b488.jpg)

![minimize-menu-closed](https://user-images.githubusercontent.com/46188345/61550353-82cca280-aa52-11e9-872f-39deacb40840.jpg)

Then I tried to move to the below the collapse button:

![minimize-menu-bottom](https://user-images.githubusercontent.com/46188345/61550526-f969a000-aa52-11e9-9f75-bc3ccd1221a2.jpg)

![minimize-menu-bottom-closed](https://user-images.githubusercontent.com/46188345/61550534-fff81780-aa52-11e9-980e-a10c3c0c6cc8.jpg)

**Which one do you think that could be the best solution?**

**Note:** when you log in with a user whit role of Admin, there is a select for change the language, at the bottom of the sidebar; as you can see by the screenshots above, I choose to hide it when the navbar is collapsed

seand7565 commented:
That looks fantastic - thanks @mfrecchiami ! I like the button being inside the collapsible panel, so it doesn't affect the layout of the rest of the page.

mfrecchiami commented:
Hi @seand7565 !

I did my job starting from your repo 💪 
I've made my branch, did the rebase with `master` and added my commits
I believed to find your repo in the _compare_ step, but it wasn't 😞 (still don't know why)
So I made another PR related to yours: https://github.com/solidusio/solidus/pull/3322

Do you want to do it in a different way? Let me know!

seand7565 commented:
@mfrecchiami No, that's perfect - thanks! I'll close this PR in favor of yours.
